diff.go

```diff
@@ -1,155 +1,416 @@
 package userconfig
 
 import (
-	"reflect"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
 )
 
 type DiffType string
 
 const (
-	// Used to indicate that the app-name in the two files was changed
-	InfoAppNameChanged DiffType = "infoAppNameChanged"
+	// DiffInfoServiceNameUpdated
+	DiffInfoServiceNameUpdated DiffType = "service-name-updated"
 
-	// Used to indicate that a node (service or component) was added. `Name` in the DiffInfo describes the node that was added.
-	InfoNodeAdded DiffType = "node-added"
+	// DiffInfoComponentAdded
+	DiffInfoComponentAdded DiffType = "component-added"
 
-	// Used to indicate that a node (service or component) was removed. `Name` in the DiffInfo describes the node that was removed.
-	InfoNodeRemoved DiffType = "node-removed"
+	// DiffInfoComponentRemoved
+	DiffInfoComponentRemoved DiffType = "component-removed"
 
-	// Used to indicate that the scaling config of a component changed.
-	InfoComponentScalingUpdated DiffType = "component-scaling-changed"
+	// DiffInfoComponentUpdated
+	DiffInfoComponentUpdated DiffType = "component-updated"
 
-	// Used to indicate that the InstanceConfig in the ComponentConfig identified by `Name` changed.
-	InfoInstanceConfigUpdated DiffType = "instance-update"
+	// DiffInfoComponentImageUpdated
+	DiffInfoComponentImageUpdated DiffType = "component-image-updated"
 
-	// Used to indicate that the component itself changed (but not the underlying InstanceConfig). `Name` in the DiffInfo describes the node that was changed.
-	InfoComponentUpdated DiffType = "component-update"
+	// DiffInfoComponentEntrypointUpdated
+	DiffInfoComponentEntrypointUpdated DiffType = "component-entrypoint-updated"
+
+	// DiffInfoComponentPortsUpdated
+	DiffInfoComponentPortsUpdated DiffType = "component-ports-updated"
+
+	// DiffInfoComponentEnvUpdated
+	DiffInfoComponentEnvUpdated DiffType = "component-env-updated"
+
+	// DiffInfoComponentVolumesUpdated
+	DiffInfoComponentVolumesUpdated DiffType = "component-volumes-updated"
+
+	// DiffInfoComponentArgsUpdated
+	DiffInfoComponentArgsUpdated DiffType = "component-args-updated"
+
+	// DiffInfoComponentDomainsUpdated
+	DiffInfoComponentDomainsUpdated DiffType = "component-domains-updated"
+
+	// DiffInfoComponentLinksUpdated
+	DiffInfoComponentLinksUpdated DiffType = "component-links-updated"
+
+	// DiffInfoComponentExposeUpdated
+	DiffInfoComponentExposeUpdated DiffType = "component-expose-updated"
+
+	// DiffInfoComponentScaleUpdated
+	DiffInfoComponentScaleUpdated DiffType = "component-scale-updated"
+
+	// DiffInfoComponentPodUpdated
+	DiffInfoComponentPodUpdated DiffType = "component-pod-updated"
+
+	// DiffInfoComponentSignalReadyUpdated
+	DiffInfoComponentSignalReadyUpdated DiffType = "component-signal-ready-updated"
 )
 
 type DiffInfo struct {
-	// What type changed: app, service, component
 	Type DiffType
 
-	// Path to the node that changed.
-	Name []string
+	Old string
+	New string
+}
+
+func (di DiffInfo) Action() string {
```

Comment:
Please add comment describing the meaning and expected use of the return values.